puffin/src/scope_details.rs

```diff
@@ -0,0 +1,212 @@
+        self
+    }
+
+    pub fn with_file(mut self, file: Cow<'static, str>) -> Self {
+        self.file_name = file;
+        self
+    }
+
+    pub fn with_line_nr(mut self, line_nr: u32) -> Self {
+        self.line_nr = line_nr;
+        self
+    }
+}
+
+#[derive(Debug, Default, Clone, PartialEq, Hash, PartialOrd, Ord, Eq)]
+/// This struct contains scope details and can be read by external libraries.
+pub struct ScopeDetailsOwned {
+    pub scope_name: Cow<'static, str>,
+    /// Shorter variant of the raw function name.
+    /// This is more descriptive and shorter.
+    /// In the case of custom scopes provided scopes this contains the `scope name`.
+    pub dynamic_function_name: Cow<'static, str>,
+    /// Raw function name with the entire type path.
+    /// This is empty for custom scopes.
+    pub raw_function_name: &'static str,
+    /// Shorter variant of the raw file path.
+    /// This is cleaned up and made consistent across platforms.
+    /// In the case of custom scopes provided scopes this contains the `file name`.
+    pub dynamic_file_path: Cow<'static, str>,
+    /// The full raw file path to the file in which the scope was located.
+    /// This is empty for custom scopes.
+    pub raw_file_path: &'static str,
+    /// The line number containing the profiling
+    pub line_nr: u32,
+    /// File name plus line number: path:number
+    pub location: String,
+}
+
+impl From<ScopeDetailsRef> for ScopeDetailsOwned {
+    fn from(value: ScopeDetailsRef) -> Self {
+        let cleaned_function_name = clean_function_name(value.raw_function_name);
+
+        ScopeDetailsOwned {
+            scope_name: format!("{cleaned_function_name}:{}", value.line_nr).into(),
+            location: format!("{cleaned_function_name}:{}", value.line_nr),
+            raw_function_name: value.raw_function_name,
+            dynamic_function_name: Cow::Owned(cleaned_function_name),
+            raw_file_path: value.file,
+            dynamic_file_path: Cow::Owned(short_file_name(value.file)),
+            line_nr: value.line_nr,
+        }
+    }
+}
+
+impl From<&CustomScopeDetails> for ScopeDetailsOwned {
+    fn from(value: &CustomScopeDetails) -> Self {
+        let mut location = String::new();
+
+        if !value.file_name.is_empty() {
+            location = format!("{}:{}", value.file_name, value.line_nr);
+        }
+
+        ScopeDetailsOwned {
+            scope_name: value.scope_name.clone(),
+            dynamic_function_name: value.function_name.clone(),
+            raw_function_name: "-", // user provided scopes are non-static
+            dynamic_file_path: value.file_name.clone(),
+            raw_file_path: "-", // user provided scopes are non-static
+            line_nr: 0,
+            location,
+        }
+    }
+}
+
+/// Scope details that can be serialized.
+#[cfg_attr(
+    feature = "serialization",
+    derive(serde::Serialize, serde::Deserialize)
+)]
+pub(crate) struct SerdeScopeDetails {
```

Comment:
This is the fourth copy now - this seems silly. Surely we can unify a bit!

For instance, we could design this so that `ScopeDetailsRef` can be converted to `CustomScopeDetails`, and then implement serde on `CustomScopeDetails`